dungeon.ml

```diff
@@ -3,22 +3,29 @@ open Library
 open Sys
 
 let ()=
-  let parse_enemy (j:json)=match j with
-    | `Assoc obj -> let e=(new enemy) in begin List.iter (fun (a:string*json) -> match a with
-        | (k,`String v) when k="name" -> e#set_name v
-        | (k,`Int i) when k="health" -> e#set_health i
-        | (k,`Int i) when k="damage" -> e#set_dmg i
+  let parse_enemy = function
+    | `Assoc obj -> List.fold_left (fun e -> function
+        | ("name", `String name)  -> Enemy.{e with name}
+        | ("health", `Int health) -> {e with health}
+        | ("damage", `Int dmg)    -> {e with dmg}
```

Comment:
`let parse_enemy j = match j with` can be written as `let parse_enemy = function`; the `function` keyword is short-hand for `let f j = match j with  (* function body *) in f`

Instead of iterating over the list and assigning the parsed attributes, we use use `List.fold_left` to [fold](https://www.reddit.com/r/ocaml/comments/5awblv/list_left_fold_how_does_it_work/) over the `obj` list.

OCaml supports matching on the native constructors like literal `string` values, `char`s, numbers, and variants:
```ocaml
match x with
| k, _ when k = "foo" ->
(* is equivalent to *)
match with 
| "foo", _ -> 
```

On the right-hand side of the matches we return the record instead.
`{x with name = v }` is short-hand for `let y = { name = v ; health = x.name ; dmg = x.name }`; it gives you a new `record` with only the entries mentioned after `with` altered (the rest are copied).
`{x with name}` is short-hand for `{x with name = name}`.